lib/internal/streams/web.js

```diff
@@ -0,0 +1,4632 @@
+          port.close();
+          break;
+      }
+    };
+
+    port.onmessageerror = () => {
+      const error = new DOMException(
+        'Internal transfered ReadableStream error',
+        'DataCloneError');
+      port.postMesssage({ type: 'error', value: error });
+      readableStreamDefaultControllerError(
+        this[kState].controller,
+        error);
+      port.close();
+    };
+  }
+
+  start(controller) {
+    this[kState].controller = controller;
+  }
+
+  async pull() {
+    this[kState].port.postMessage({ type: 'pull' });
+  }
+
+  async cancel(reason) {
+    try {
+      this[kState].port.postMessage({ type: 'error', value: reason });
+    } catch (error) {
+      this[kState].port.postMessage({ type: 'error', value: error });
+      throw error;
+    } finally {
+      this[kState].port.close();
+    }
+  }
+}
+
+class CrossRealmTransformWritableSink {
+  constructor(port) {
+    this[kState] = {
+      port,
+      controller: undefined,
+      backpressurePromise: createDeferredPromise(),
+    };
+
+    port.onmessage = ({ data }) => {
+      assert(typeof data === 'object');
+      const {
+        type,
+        value
+      } = { ...data };
+      assert(typeof type === 'string');
+      switch (type) {
+        case 'pull':
+          if (this[kState].backpressurePromise !== undefined)
+            this[kState].backpressurePromise.resolve?.();
+          this[kState].backpressurePromise = undefined;
+          break;
+        case 'error':
+          writableStreamDefaultControllerErrorIfNeeded(
+            this[kState].controller,
+            value);
+          if (this[kState].backpressurePromise !== undefined)
+            this[kState].backpressurePromise.resolve?.();
+          this[kState].backpressurePromise = undefined;
+          break;
+      }
+    };
+    port.onmessageerror = () => {
+      const error = new DOMException(
+        'Internal transfered ReadableStream error',
+        'DataCloneError');
+      port.postMesssage({ type: 'error', value: error });
+      writableStreamDefaultControllerErrorIfNeeded(
+        this[kState].controller,
+        error);
+      port.close();
+    };
+
+  }
+
+  start(controller) {
+    this[kState].controller = controller;
+  }
+
+  async write(chunk) {
+    if (this[kState].backpressurePromise === undefined) {
+      this[kState].backpressurePromise = {
+        promise: PromiseResolve(),
+        resolve: undefined,
+        reject: undefined,
+      };
+    }
+    await this[kState].backpressurePromise.promise;
+    this[kState].backpressurePromise = createDeferredPromise();
+    try {
+      this[kState].port.postMessage({ type: 'chunk', value: chunk });
+    } catch (error) {
+      this[kState].port.postMessage({ type: 'error', value: error });
+      this[kState].port.close();
+      throw error;
+    }
+  }
+
+  async close() {
+    this[kState].port.postMessage({ type: 'close' });
+    this[kState].port.close();
+  }
```

Comment:
nothing's awaited or returned here - so this will always be an immediately resolved promise for undefined. is this intentional?